# Students' Dropout and Academic success Analysis

**Цель**: Разработать систему раннего выявления студентов группы риска (риск отчисления) на ранних этапах обучения.

###  Exploratory Data Analysis

- Загрузка данных, изучение общей структуры данных

In [1]:
import pandas as pd

In [2]:
# Загрузка данных

In [3]:
df.head(10)

,Marital status,Application mode,Application order,Course,Daytime/evening attendance\t,Previous qualification,Previous qualification (grade),Nacionality,Mother's qualification,Father's qualification,...,Curricular units 2nd sem (credited),Curricular units 2nd sem (enrolled),Curricular units 2nd sem (evaluations),Curricular units 2nd sem (approved),Curricular units 2nd sem (grade),Curricular units 2nd sem (without evaluations),Unemployment rate,Inflation rate,GDP,Target
0,1,17,5,171,1,1,122.0,1,19,12,...,0,0,0,0,0.000000,0,10.8,1.4,1.74,Dropout
1,1,15,1,9254,1,1,160.0,1,1,3,...,0,6,6,6,13.666667,0,13.9,-0.3,0.79,Graduate
2,1,1,5,9070,1,1,122.0,1,37,37,...,0,6,0,0,0.000000,0,10.8,1.4,1.74,Dropout
3,1,17,2,9773,1,1,122.0,1,38,37,...,0,6,10,5,12.400000,0,9.4,-0.8,-3.12,Graduate
4,2,39,1,8014,0,1,100.0,1,37,38,...,0,6,6,6,13.000000,0,13.9,-0.3,0.79,Graduate
5,2,39,1,9991,0,19,133.1,1,37,37,...,0,5,17,5,11.500000,5,16.2,0.3,-0.92,Graduate
6,1,1,1,9500,1,1,142.0,1,19,38,...,0,8,8,8,14.345000,0,15.5,2.8,-4.06,Graduate
7,1,18,4,9254,1,1,119.0,1,37,37,...,0,5,5,0,0.000000,0,15.5,2.8,-4.06,Dropout
8,1,1,3,9238,1,1,137.0,62,1,1,...,0,6,7,6,14.142857,0,16.2,0.3,-0.92,Graduate
9,1,1,1,9238,1,1,138.0,1,1,19,...,0,6,14,2,13.500000,0,8.9,1.4,3.51,Dropout


In [4]:
# Размерность (строки, столбцы)
df.shape  

(4424, 37)

In [5]:
# Проверка данных на пропуски и соответствие форматов данных
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 37 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Marital status                                  4424 non-null   int64  
 1   Application mode                                4424 non-null   int64  
 2   Application order                               4424 non-null   int64  
 3   Course                                          4424 non-null   int64  
 4   Daytime/evening attendance	                     4424 non-null   int64  
 5   Previous qualification                          4424 non-null   int64  
 6   Previous qualification (grade)                  4424 non-null   float64
 7   Nacionality                                     4424 non-null   int64  
 8   Mother's qualification                          4424 non-null   int64  
 9   Father's qualification                   

In [6]:
# Для удобства работы с данными переименуем столбцы, приведя их к snake case и сократив по возможности

# Создаем словарь для переименования столбцов 
column_rename = {
    'Marital status': 'marital_status',
    'Application mode': 'app_mode',
    'Application order': 'app_order',
    'Course': 'course',
    'Daytime/evening attendance\t': 'attendance',
    'Previous qualification': 'prev_edu',
    'Previous qualification (grade)': 'prev_grade',
    'Nacionality': 'nationality',
    "Mother's qualification": 'mom_edu',
    "Father's qualification": 'dad_edu',
    "Mother's occupation": 'mom_job',
    "Father's occupation": 'dad_job',
    'Admission grade': 'admission_grade',
    'Displaced': 'displaced',
    'Educational special needs': 'special_needs',
    'Debtor': 'debtor',
    'Tuition fees up to date': 'fees_paid',
    'Gender': 'gender',
    'Scholarship holder': 'scholarship',
    'Age at enrollment': 'age',
    'International': 'international',
    'Curricular units 1st sem (credited)': 'sem1_credited',
    'Curricular units 1st sem (enrolled)': 'sem1_enrolled',
    'Curricular units 1st sem (evaluations)': 'sem1_evaluations',
    'Curricular units 1st sem (approved)': 'sem1_passed',
    'Curricular units 1st sem (grade)': 'sem1_grade',  
    'Curricular units 1st sem (without evaluations)': 'sem1_no_grade',
    'Curricular units 2nd sem (credited)': 'sem2_credited',
    'Curricular units 2nd sem (enrolled)': 'sem2_enrolled',
    'Curricular units 2nd sem (evaluations)': 'sem2_evaluations',
    'Curricular units 2nd sem (approved)': 'sem2_passed',
    'Curricular units 2nd sem (grade)': 'sem2_grade',
    'Curricular units 2nd sem (without evaluations)': 'sem2_no_grade',
    'Unemployment rate': 'unemployment',
    'Inflation rate': 'inflation',
    'GDP': 'gdp',
    'Target': 'target'
}

# Применяем переименование
df = df.rename(columns=column_rename)
df.columns

Index(['marital_status', 'app_mode', 'app_order', 'course', 'attendance',
       'prev_edu', 'prev_grade', 'nationality', 'mom_edu', 'dad_edu',
       'mom_job', 'dad_job', 'admission_grade', 'displaced', 'special_needs',
       'debtor', 'fees_paid', 'gender', 'scholarship', 'age', 'international',
       'sem1_credited', 'sem1_enrolled', 'sem1_evaluations', 'sem1_passed',
       'sem1_grade', 'sem1_no_grade', 'sem2_credited', 'sem2_enrolled',
       'sem2_evaluations', 'sem2_passed', 'sem2_grade', 'sem2_no_grade',
       'unemployment', 'inflation', 'gdp', 'target'],
      dtype='object')

In [7]:
 # Сводка по числовым признакам
df.describe()  

,marital_status,app_mode,app_order,course,attendance,prev_edu,prev_grade,nationality,mom_edu,dad_edu,...,sem1_no_grade,sem2_credited,sem2_enrolled,sem2_evaluations,sem2_passed,sem2_grade,sem2_no_grade,unemployment,inflation,gdp
count,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,...,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000
mean,1.178571,18.669078,1.727848,8856.642631,0.890823,4.577758,132.613314,1.873192,19.561935,22.275316,...,0.137658,0.541817,6.232143,8.063291,4.435805,10.230206,0.150316,11.566139,1.228029,0.001969
std,0.605747,17.484682,1.313793,2063.566416,0.311897,10.216592,13.188332,6.914514,15.603186,15.343108,...,0.690880,1.918546,2.195951,3.947951,3.014764,5.210808,0.753774,2.663850,1.382711,2.269935
min,1.000000,1.000000,0.000000,33.000000,0.000000,1.000000,95.000000,1.000000,1.000000,1.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,7.600000,-0.800000,-4.060000
25%,1.000000,1.000000,1.000000,9085.000000,1.000000,1.000000,125.000000,1.000000,2.000000,3.000000,...,0.000000,0.000000,5.000000,6.000000,2.000000,10.750000,0.000000,9.400000,0.300000,-1.700000
50%,1.000000,17.000000,1.000000,9238.000000,1.000000,1.000000,133.100000,1.000000,19.000000,19.000000,...,0.000000,0.000000,6.000000,8.000000,5.000000,12.200000,0.000000,11.100000,1.400000,0.320000
75%,1.000000,39.000000,2.000000,9556.000000,1.000000,1.000000,140.000000,1.000000,37.000000,37.000000,...,0.000000,0.000000,7.000000,10.000000,6.000000,13.333333,0.000000,13.900000,2.600000,1.790000
max,6.000000,57.000000,9.000000,9991.000000,1.000000,43.000000,190.000000,109.000000,44.000000,44.000000,...,12.000000,19.000000,23.000000,33.000000,20.000000,18.571429,12.000000,16.200000,3.700000,3.510000


In [8]:
# Сводка по категориальным признакам
df.describe(include=['object'])

,target
count,4424
unique,3
top,Graduate
freq,2209


In [9]:
# Уникальные значения в каждом столбце
df.nunique()

marital_status        6
app_mode             18
app_order             8
course               17
attendance            2
prev_edu             17
prev_grade          101
nationality          21
mom_edu              29
dad_edu              34
mom_job              32
dad_job              46
admission_grade     620
displaced             2
special_needs         2
debtor                2
fees_paid             2
gender                2
scholarship           2
age                  46
international         2
sem1_credited        21
sem1_enrolled        23
sem1_evaluations     35
sem1_passed          23
sem1_grade          805
sem1_no_grade        11
sem2_credited        19
sem2_enrolled        22
sem2_evaluations     30
sem2_passed          20
sem2_grade          786
sem2_no_grade        10
unemployment         10
inflation             9
gdp                  10
target                3
dtype: int64

-  Анализ целевой переменной (Target)

In [10]:
# Распределение студентов отчислившихся, выпустившихся и продолжающих обучение
df['target'].value_counts()

target
Graduate    2209
Dropout     1421
Enrolled     794
Name: count, dtype: int64

In [11]:
# Распределение студентов отчислившихся, выпустившихся и продолжающих обучение(в %)
df['target'].value_counts(normalize=True) * 100

target
Graduate    49.932188
Dropout     32.120253
Enrolled    17.947559
Name: proportion, dtype: float64

- Статистика по демографическим признакам

In [12]:
# Основные метрики
print("ДЕМОГРАФИЯ - ОСНОВНЫЕ МЕТРИКИ:")
print(f"Всего студентов: {len(df)}")
print(f"Средний возраст: {df['age'].mean():.1f} лет")
print(f"Медианный возраст: {df['age'].median():.0f} лет")

male_count = df['gender'].sum()  
female_count = len(df) - male_count  
print(f"\nМужчины: {male_count} ({male_count/len(df)*100:.1f}%)")
print(f"Женщины: {female_count} ({female_count/len(df)*100:.1f}%)")

international_count = df['international'].sum()
local_count = len(df) - international_count
print(f"Резиденство: {local_count} местных/{international_count} иностранных "
      f"({local_count/len(df)*100:.1f}%/{international_count/len(df)*100:.1f}%)")
print('-------------')

# Создаем читаемые метки
df['gender_label'] = df['gender'].map({0: 'Female', 1: 'Male'})
df['international_label'] = df['international'].map({0: 'Local', 1: 'International'})

# Таблица успешности
success_table = pd.pivot_table(df, 
                               index=['gender_label', 'international_label'],
                               columns='target',
                               aggfunc='size',
                               fill_value=0)
print("\nУспешность по демографическим группам:")
success_table

ДЕМОГРАФИЯ - ОСНОВНЫЕ МЕТРИКИ:
Всего студентов: 4424
Средний возраст: 23.3 лет
Медианный возраст: 20 лет

Мужчины: 1556 (35.2%)
Женщины: 2868 (64.8%)
Резиденство: 4314 местных/110 иностранных (97.5%/2.5%)
-------------

Успешность по демографическим группам:


target                            Dropout  Enrolled  Graduate
gender_label international_label                             
Female       International             23        15        42
             Local                    697       472      1619
Male         International              9         9        12
             Local                    692       298       536

In [13]:
# Сохраняем полученный датафрейм
df.to_csv("C:/Users/stargirl/PETproj/Новая папка/data_cleaned.csv")

**Промежуточный вывод:** 

Датасет не содержит пропусков, форматы данных верны. Датасет достаточно большого размера для дальнейшего анализа.

Анализ целевой переменной показал умеренный дисбаланс классов, но не критичный для анализа.

Средний возраст среди студентов 23 года, медианный - 20 лет. Разница между значениями небольшая и указывает на наличие "хвоста" среди более взрослых студентов.

Гендерный дисбаланс: в выборке значительно больше женщин, чем мужчин. Женщины (местные и иностранки) показывают лучшие результаты, чем мужчины.

Иностранные студенты составляют меньшинство (суммарно 2.5% от всех студентов).

**Главный инсайт:** Женщины-местные студентки — самая многочисленная (1619 человек) и успешная группа, формирующая позитивную статистику по выпускникам в целом.

**Рекомендация:** Университету стоит изучить причины более низкой успеваемости среди мужчин и рассмотреть меры поддержки для этой группы, особенно для местных студентов-мужчин, где самый высокий процент отчислений.